lib/smack/IntegerOverflowChecker.cpp

```diff
@@ -0,0 +1,159 @@
+  }
+}
+
+Value* IntegerOverflowChecker::mkResult(Value* v, int bits, Instruction* i, bool check) {
+  if (check)
+    return CastInst::CreateTruncOrBitCast(v, IntegerType::get(i->getFunction()->getContext(), bits), "", i);
+  else
+    return v;
+}
+
+void IntegerOverflowChecker::mkCheck(Function* co, BinaryOperator* flag, Instruction* i) {
+  ArrayRef<Value*> args(CastInst::CreateIntegerCast(flag, co->arg_begin()->getType(), false, "", i));
+  CallInst::Create(co, args, "", i);
+}
+
+void IntegerOverflowChecker::mkBlockingAssume(Function* va, BinaryOperator* flag, Instruction* i) {
+  ArrayRef<Value*> args(CastInst::CreateIntegerCast(BinaryOperator::CreateNot(flag, "", i),
+        va->arg_begin()->getType(), false, "", i));
+  CallInst::Create(va, args, "", i);
+}
+
+bool IntegerOverflowChecker::runOnModule(Module& m) {
+  Function* co = m.getFunction("__SMACK_check_overflow");
+  assert(co != NULL && "Function __SMACK_check_overflow should be present.");
+  Function* va = m.getFunction("__VERIFIER_assume");
+  assert(va != NULL && "Function __VERIFIER_assume should be present.");
+  std::vector<Instruction*> instToRemove;
+  for (auto& F : m) {
+    if (!Naming::isSmackName(F.getName())) {
+      for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+        if (auto ei = dyn_cast<ExtractValueInst>(&*I)) {
+          if (auto ci = dyn_cast<CallInst>(ei->getAggregateOperand())) {
+            Function* f = ci->getCalledFunction();
+            SmallVectorImpl<StringRef> *ar = new SmallVector<StringRef, 3>;
+            if (f && f->hasName() && OVERFLOW_INTRINSICS.match(f->getName().str(), ar)) {
+              if (ei->getIndices()[0] == 1) {
+                Instruction* prev = &*std::prev(I);
+                bool isSigned = ar->begin()[1].str() == "s";
+                std::string op = ar->begin()[2].str();
+                std::string len = ar->begin()[3].str();
+                int bits = std::stoi(len);
+                Value* eo1 = extValue(ci->getArgOperand(0), bits, isSigned, &*I, SmackOptions::IntegerOverflow);
+                Value* eo2 = extValue(ci->getArgOperand(1), bits, isSigned, &*I, SmackOptions::IntegerOverflow);
+                BinaryOperator* ai = BinaryOperator::Create(INSTRUCTION_TABLE.at(op), eo1, eo2, "", &*I);
+                if (prev && isa<ExtractValueInst>(prev)) {
+                  ExtractValueInst* pei = cast<ExtractValueInst>(prev);
+                  if (auto pci = dyn_cast<CallInst>(pei->getAggregateOperand())) {
+                    if (pci == ci) {
+                      Value* r = mkResult(ai, bits, &*I, SmackOptions::IntegerOverflow);
+                      prev->replaceAllUsesWith(r);
+                      instToRemove.push_back(prev);
+                    }
+                  }
+                }
+                BinaryOperator* flag = mkFlag(ai, bits, isSigned, &*I, SmackOptions::IntegerOverflow);
+                mkCheck(co, flag, &*I);
```

Comment:
Ah, I see now. So maybe we should call this one `addCheck`.

Reply:
Done.